pkg/networkserver/grpc_asns.go

```diff
@@ -49,51 +59,64 @@ func (ns *NetworkServer) LinkApplication(link ttnpb.AsNs_LinkApplicationServer)
 	ids := ttnpb.ApplicationIdentifiers{
 		ApplicationID: rpcmetadata.FromIncomingContext(ctx).ID,
 	}
-	var err error
-	if err = ids.ValidateContext(ctx); err != nil {
+	if err := ids.ValidateContext(ctx); err != nil {
 		return err
 	}
-	if err = rights.RequireApplication(ctx, ids, ttnpb.RIGHT_APPLICATION_LINK); err != nil {
+	if err := rights.RequireApplication(ctx, ids, ttnpb.RIGHT_APPLICATION_LINK); err != nil {
 		return err
 	}
 
 	ws := &applicationUpStream{
-		AsNs_LinkApplicationServer: link,
-		closeCh:                    make(chan struct{}),
+		closeCh: make(chan struct{}),
 	}
 	defer func() {
 		close(ws.closeCh)
 	}()
 
 	uid := unique.ID(ctx, ids)
-
 	logger := log.FromContext(ctx).WithField("application_uid", uid)
 
 	v, ok := ns.applicationServers.LoadOrStore(uid, ws)
 	for ok {
-		logger.Debug("Close existing link")
+		logger.Debug("Close existing application link")
 		if err := v.(io.Closer).Close(); err != nil {
-			logger.WithError(err).Warn("Failed to close existing link")
+			logger.WithError(err).Warn("Failed to close existing application link")
 		}
 		v, ok = ns.applicationServers.LoadOrStore(uid, ws)
 	}
 	defer ns.applicationServers.Delete(uid)
 
 	logger.Debug("Linked application")
-
 	events.Publish(evtBeginApplicationLink(ctx, ids, nil))
 
-	select {
-	case <-ctx.Done():
-		err = ctx.Err()
-		logger.WithError(err).Debug("Context done - close stream")
-		events.Publish(evtEndApplicationLink(ctx, ids, err))
-		return err
+	ctx, cancel := context.WithCancel(ctx)
+	go func() {
+		<-ws.closeCh
+		cancel()
+	}()
+	for {
+		if err := ns.applicationUplinks.Range(ctx, ids, func(ctx context.Context, up *ttnpb.ApplicationUp) error {
+			select {
+			case <-ctx.Done():
+				return ctx.Err()
+
+			case <-ws.closeCh:
+				cancel()
+				<-ctx.Done()
+				return ctx.Err()
 
-	case ws.closeCh <- struct{}{}:
-		logger.Debug("Link closed - close stream")
-		events.Publish(evtEndApplicationLink(ctx, ids, nil))
-		return nil
+			default:
+				if err := link.Send(up); err != nil {
+					return err
+				}
+				_, err := link.Recv()
+				return err
+			}
+		}); err != nil {
+			logger.WithError(err).Debug("Close application link")
+			events.Publish(evtEndApplicationLink(ctx, ids, err))
```

Comment:
It *can* be, but there's not reason to add such complexity here.
`LinkApplication` only returns when `Range` errors, hence that is also the only time when `EndApplicationLink` will be published and the payload is the error returned by `Range`

Reply:
OK, I see now.

Why aren't we working with a channel between the registry and here? What is the upside of blocking on a read manually vs using a channel?

Reply:
1. Using channel here is inconsistent with other similar functionality and e.g. `sync.Map.Range` in stdlib
2. Channel requires an additional goroutine
3. Channel requires additional complexity(a wrapper struct is needed to hold context and uplink(s))
4. There's no reason to use a channel here, since we have to wait for `Range` to return completely anyway

What do you mean by "blocking manually"? `Range` blocks forever until `<-ctx.Done()` succeeds or error occurs

Reply:
> 1. Using channel here is inconsistent with other similar functionality and e.g. `sync.Map.Range` in stdlib

That one, as well as range implementations of existing registries, are about existing items. This is a blocking consumer which is fundamentally different.

> 2. Channel requires an additional goroutine

Where?

Also, are you considering one goroutine per application link too expensive overhead?

> 3. Channel requires additional complexity - a wrapper struct is needed to hold context and uplink(s) and additional logic both here as well as in the interface implementations.

Yes, see also `ContextualApplicationUp` in Application Server.

> 4. There's no reason to use a channel here, since we have to wait for `Range` to return completely anyway

Why?

---

How about `Consume(..., callback)`, drop the `for { }`, and let the implementation continuously call the callback on new items?

Still, I think a channel (with contextual items), is cleaner.

Reply:
> > 1. Using channel here is inconsistent with other similar functionality and e.g. `sync.Map.Range` in stdlib

> That one, as well as range implementations of existing registries, are about existing items. This is a blocking consumer which is fundamentally different.
Why so? The only difference from e.g. `sync.Map.Range`, here is that `sync.Map` is finite, whereas we're ranging over an infinite data structure.
This is modeled after `for x := range <channel>`.

> > 2. Channel requires an additional goroutine

> Where?

> Also, are you considering one goroutine per application link too expensive overhead?

There are different ways to "use a channel" here, if you provide me the method signature you are thinking about, I will provide you with the details.
Assuming the signature is `Range(<args>) (<-chan W, error)`, where `W` contains `*ttnpb.ApplicationUp`, `Context` and `error`, Redis implementation will have to `XReadGroup` in a goroutine in order to send downlinks on the channel.

a goroutine takes at least 2KB on Go 1.13
I think that if we can avoid using N*2KB extra memory space, where N is potentially quite large(N=application count in worst case), we should.
My biggest issue with having a goroutine is the added complexity of handling it. We would need additional `select` statements as well.

> > 4. There's no reason to use a channel here, since we have to wait for `Range` to return completely anyway

> Why?

https://github.com/TheThingsNetwork/lorawan-stack/pull/1524/files#diff-8f540de6ab7a805994f024884bbfc1e8R73

The simplicity of current solution is due to this assumption.



> How about Consume(..., callback), drop the for { }, and let the implementation continuously call the callback on new items?

> Still, I think a channel (with contextual items), is cleaner.

So current Redis implementation does *continuously call the callback on new items*, as well as the failed-to-process items. `for {}` block is not strictly necessary, it's there for 2 reasons:
1. Possible other implementations of `ApplicationUplinkQueue` may be represented as a finite data structure, in which case `Range` would return once the end is reached
2. Later we will add failure recovery and/or retrying of failed tasks, where we will have to call `Range` again.


Also, please note, that the actual callback need not be an anonymous function, and it may be configurable and have different implementations(e.g. `ns.applicationUplinks.Range(ctx, ids, ns.asUplinkHandler)`.
Also, the
```
			select {
			case <-ctx.Done():
				return ctx.Err()

			case <-ws.closeCh:
				cancel()
				<-ctx.Done()
				return ctx.Err()

			default:
```
block is, in fact, optional and is there just there to ensure failing quick and minimizing the chance of sending the uplink on the link after the context is done and/or link is closed - this is handled in goroutines, so there are no guarantees on the actual execution time, but it should be enough.
So, in fact the whole `Range` call can just look like this:
```
		if err := ns.applicationUplinks.Range(ctx, ids, func(ctx context.Context, up *ttnpb.ApplicationUp) error {
			if err := link.Send(up); err != nil {
				return err
			}
			_, err := link.Recv()
			return err
		}); err != nil {
			logger.WithError(err).Debug("Close application link")
			events.Publish(evtEndApplicationLink(ctx, ids, err))
			return err
		}
```


> Still, I think a channel (with contextual items), is cleaner.

Can you describe what you find "not clean" about current implementation? Channel-based implementation seems to be way more "dirty" to me in fact.

Reply:
> Assuming the signature is `Range(<args>) (<-chan W, error)`, where `W` contains `*ttnpb.ApplicationUp`, `Context` and `error`, Redis implementation will have to `XReadGroup` in a goroutine in order to send downlinks on the channel.
> 
> a goroutine takes at least 2KB on Go 1.13
> I think that if we can avoid using N*2KB extra memory space, where N is potentially quite large(N=application count in worst case), we should.
> My biggest issue with having a goroutine is the added complexity of handling it. We would need additional `select` statements as well.

I care less about memory in this case and I don't see the complexity compared to what it looks like now.

> The simplicity of current solution is due to this assumption.

This is implementation-specific and not part of the interface.

> So current Redis implementation does _continuously call the callback on new items_, as well as the failed-to-process items

This is not defined in the interface either.

Also why is it called `Range()` and not `Consume()` already?

> 2\. Later we will add failure recovery and/or retrying of failed tasks, where we will have to call `Range` again.

OK

So basically `Range()` is the wrong term; `Consume()` or `Subscribe()` make more sense to me

Reply:
Please see the updated version.